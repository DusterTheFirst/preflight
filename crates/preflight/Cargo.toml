[package]
name = "preflight"
version = "0.1.0"
authors = ["Zachary Kohnen <14093962+DusterTheFirst@users.noreply.github.com>"]
edition = "2018"

[dependencies]
termcolor = "1.1"
timescale = { path = "../timescale" }
cargo_metadata = "0.12"
thiserror = "1.0"
structopt = "0.3"
dlopen = "0.1"
dlopen_derive = "0.1"
preflight_impl = { path = "./impl" }
