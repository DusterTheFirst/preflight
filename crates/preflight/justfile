preflight-example action:
    cargo build --bin cargo-preflight
    cd example && cargo preflight {{action}}