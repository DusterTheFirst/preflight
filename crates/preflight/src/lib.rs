#[macro_use]
extern crate dlopen_derive;

pub mod args;
pub mod cargo;
pub mod shell;
pub mod api;

pub use preflight_impl::*;