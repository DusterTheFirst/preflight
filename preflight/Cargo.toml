[package]
name = "cargo-preflight"
version = "0.1.0"
authors = ["Zachary Kohnen <14093962+DusterTheFirst@users.noreply.github.com>"]
edition = "2018"

[dependencies]
anyhow = "1.0"
cargo_metadata = "0.12"
dlopen = "0.1"
dlopen_derive = "0.1"
glob = "0.3"
indoc = "1.0"
lazy_static = "1.4"
open = "1.4"
preflight_impl = { path = "./impl" }
structopt = "0.3"
termcolor = "1.1"
textwrap = "0.13"
timescale = { path = "../timescale" }
uuid = { version = "0.8", features = ["v4"] }
