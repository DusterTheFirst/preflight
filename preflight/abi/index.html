<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `abi` mod in crate `preflight`."><meta name="keywords" content="rust, rustlang, rust-lang, abi"><title>preflight::abi - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings"></script><script src="../../storage.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg">
<link rel="alternate icon" type="image/png" href="../../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../../preflight/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><p class="location">Module abi</p><div class="sidebar-elems"><div class="block items"><ul><li><a href="#types">Type Definitions</a></li></ul></div><p class="location"><a href="../index.html">preflight</a></p><script>window.sidebarCurrent = {name: "abi", ty: "mod", relpath: "../"};</script><script defer src="../sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><span class="help-button">?</span>
                <a id="settings-menu" href="../../settings.html"><img src="../../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/preflight/abi.rs.html#1-22" title="goto source code">[src]</a></span><span class="in-band">Module <a href="../index.html">preflight</a>::<wbr><a class="mod" href="">abi</a></span></h1><div class="docblock"><p><strong>This module should never be consumed directly.</strong> The ABI used to communicate
between the preflight runner and the avionics.</p>
<p>This module's contents are only made public for use by code generated by the
<a href="../../preflight/attr.avionics_harness.html"><code>avionics_harness</code></a> macro</p>
</div><h2 id="types" class="section-header"><a href="#types">Type Definitions</a></h2>
<table><tr class="module-item"><td><a class="type" href="type.AvionicsGuide.html" title="preflight::abi::AvionicsGuide type">AvionicsGuide</a></td><td class="docblock-short"><p>The entry point into the user's avionics</p>
</td></tr><tr class="module-item"><td><a class="type" href="type.PanicCallback.html" title="preflight::abi::PanicCallback type">PanicCallback</a></td><td class="docblock-short"><p>Callback that is called when the driven avionics panics.</p>
</td></tr><tr class="module-item"><td><a class="type" href="type.SetPanicCallback.html" title="preflight::abi::SetPanicCallback type">SetPanicCallback</a></td><td class="docblock-short"><p>Function exposed to the preflight runner that can set the <a href="../../preflight/abi/type.PanicCallback.html" title="PanicCallback"><code>PanicCallback</code></a></p>
</td></tr></table></section><section id="search" class="content hidden"></section><section class="footer"></section><script>window.rootPath = "../../";window.currentCrate = "preflight";</script><script src="../../main.js"></script><script defer src="../../search-index.js"></script></body></html>