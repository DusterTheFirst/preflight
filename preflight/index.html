<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `preflight` crate."><meta name="keywords" content="rust, rustlang, rust-lang, preflight"><title>preflight - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings"></script><script src="../storage.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../favicon.svg">
<link rel="alternate icon" type="image/png" href="../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu">&#9776;</div><a href='../preflight/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><p class="location">Crate preflight</p><div class="block version"><p>Version 0.1.0</p></div><div class="sidebar-elems"><a id="all-types" href="all.html"><p>See all preflight's items</p></a><div class="block items"><ul><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#types">Type Definitions</a></li></ul></div><p class="location"></p><script>window.sidebarCurrent = {name: "preflight", ty: "mod", relpath: "../"};</script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../brush.svg" width="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><span class="help-button">?</span>
                <a id="settings-menu" href="../settings.html"><img src="../wheel.svg" width="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../src/preflight/lib.rs.html#1-161" title="goto source code">[src]</a></span><span class="in-band">Crate <a class="mod" href="">preflight</a></span></h1><div class="docblock"><p>Utilities to create and test hardware agnostic flight systems with little friction</p>
<p>This preflight crate goes hand and hand with the <code>cargo_preflight</code> cargo
subcommand. Alone, this crate will produce a stable, plugable interface for
a flight computer to link to. When paired with the <code>cargo_preflight</code> command,
users are able to run the flight systems through rigorous simulations and
tests to verify their integrity.</p>
</div><h2 id="reexports" class="section-header"><a href="#reexports">Re-exports</a></h2>
<table><tr><td><code>pub use <a class="mod" href="../uom/index.html" title="mod uom">uom</a>;</code></td></tr></table><h2 id="modules" class="section-header"><a href="#modules">Modules</a></h2>
<table><tr class="module-item"><td><a class="mod" href="abi/index.html" title="preflight::abi mod">abi</a></td><td class="docblock-short"><p><strong>This module should never be consumed directly.</strong> The ABI used to communicate
between the preflight runner and the avionics.</p>
</td></tr></table><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<table><tr class="module-item"><td><a class="struct" href="struct.Guidance.html" title="preflight::Guidance struct">Guidance</a></td><td class="docblock-short"><p>A hardware agnostic guidance signal</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.Sensors.html" title="preflight::Sensors struct">Sensors</a></td><td class="docblock-short"><p>Generic sensor values that are collected from the flight hardware which would
be useful to calculate position and velocity.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.ThrustVector.html" title="preflight::ThrustVector struct">ThrustVector</a></td><td class="docblock-short"><p>A call for thrust vectoring hardware to produce a thrust at the given
vector</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.Vector3.html" title="preflight::Vector3 struct">Vector3</a></td><td class="docblock-short"><p>A vector representing a quantity in 3 dimensional space</p>
</td></tr></table><h2 id="enums" class="section-header"><a href="#enums">Enums</a></h2>
<table><tr class="module-item"><td><a class="enum" href="enum.AbortCause.html" title="preflight::AbortCause enum">AbortCause</a></td><td class="docblock-short"><p>Underlying cause for an abort</p>
</td></tr><tr class="module-item"><td><a class="enum" href="enum.Control.html" title="preflight::Control enum">Control</a></td><td class="docblock-short"><p>Various control signals that the avionics can produce</p>
</td></tr></table><h2 id="traits" class="section-header"><a href="#traits">Traits</a></h2>
<table><tr class="module-item"><td><a class="trait" href="trait.Avionics.html" title="preflight::Avionics trait">Avionics</a></td><td class="docblock-short"><p>Hardware agnostic avionics system</p>
</td></tr></table><h2 id="types" class="section-header"><a href="#types">Type Definitions</a></h2>
<table><tr class="module-item"><td><a class="type" href="type.Quantity.html" title="preflight::Quantity type">Quantity</a></td><td class="docblock-short"><p>Generic <a href="../uom/index.html" title="uom"><code>uom</code></a> quantity using f32 as the storage type</p>
</td></tr></table><h2 id="attributes" class="section-header"><a href="#attributes">Attribute Macros</a></h2>
<table><tr class="module-item"><td><a class="attr" href="attr.avionics_harness.html" title="preflight::avionics_harness attr">avionics_harness</a></td><td class="docblock-short"><p>Harness to connect hardware agnostic flight systems to firmware or to the
<code>preflight_cargo</code> utility</p>
</td></tr></table></section><section id="search" class="content hidden"></section><section class="footer"></section><script>window.rootPath = "../";window.currentCrate = "preflight";</script><script src="../main.js"></script><script defer src="../search-index.js"></script></body></html>