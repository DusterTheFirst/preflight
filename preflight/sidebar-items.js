initSidebarItems({"attr":[["avionics_harness","Harness to connect hardware agnostic flight systems to firmware or to the `preflight_cargo` utility"]],"enum":[["AbortCause","Underlying cause for an abort"],["Control","Various control signals that the avionics can produce"]],"mod":[["abi","This module should never be consumed directly. The ABI used to communicate between the preflight runner and the avionics."]],"struct":[["Guidance","A hardware agnostic guidance signal"],["Sensors","Generic sensor values that are collected from the flight hardware which would be useful to calculate position and velocity."],["ThrustVector","A call for thrust vectoring hardware to produce a thrust at the given vector"],["Vector3","A vector representing a quantity in 3 dimensional space"]],"trait":[["Avionics","Hardware agnostic avionics system"]],"type":[["Quantity","Generic [`uom`] quantity using f32 as the storage type"]]});