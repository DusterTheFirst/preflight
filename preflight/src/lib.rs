//! Utilities to create and test hardware agnostic flight systems with little friction
//!
//! This preflight crate goes hand and hand with the [`cargo_preflight`] cargo
//! subcommand. Alone, this crate will produce a stable, plugable interface for
//! a flight computer to link to. When paired with the [`cargo_preflight`] command,
//! users are able to run the flight systems through rigorous simulations and
//! tests to verify their integrity.

#![no_std]
#![forbid(unsafe_code)]
#![warn(missing_docs)]

use core::{
    fmt::{self, Debug, Formatter},
    marker::PhantomData,
};

pub use preflight_macros::avionics_harness;
pub use uom;
use uom::si::{
    acceleration, angle, angular_velocity, length, magnetic_flux_density, Dimension, SI,
};

pub mod abi;

/// Generic [`uom`] quantity using f64 as the storage type
pub type Quantity<T> = uom::si::Quantity<T, SI<f64>, f64>;

/// Generic sensor values that are collected from the flight hardware which would
/// be useful to calculate position and velocity.
#[repr(C)]
#[derive(Debug)]
pub struct Sensors {
    /// Calculated altitude
    pub altitude: Quantity<length::Dimension>,
    /// Three axis of linear acceleration data (acceleration minus gravity) in m/s^2
    pub linear_acceleration: Vector3<acceleration::Dimension>,
    /// Three axis of gravitational acceleration (minus any movement) in m/s^2
    pub gravity_acceleration: Vector3<acceleration::Dimension>,
    /// Three axis of acceleration (gravity + linear motion) in m/s^2
    pub both_acceleration: Vector3<acceleration::Dimension>,
    /// Three axis orientation data based on a 360Â° sphere
    pub orientation: Vector3<angle::Dimension>,
    /// Three axis of 'rotation speed' in rad/s
    pub angular_velocity: Vector3<angular_velocity::Dimension>,
    /// Three axis of magnetic field sensing in micro Tesla (uT)
    pub magnetic_field: Vector3<magnetic_flux_density::Dimension>,
}

/// A vector representing a quantity in 3 dimensional space
#[repr(C)]
pub struct Vector3<T: Dimension + ?Sized> {
    x: Quantity<T>,
    y: Quantity<T>,
    z: Quantity<T>,
}

impl<T: Dimension + ?Sized> Debug for Vector3<T> {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        // if f.alternate() {

        // }
        // write!(f, "({}, {}, {})", self.x.value, self.y.value, self.z);
        f.debug_struct("Vector3")
            .field("x", &self.x) //Quantity::format_args(self.x, Abbreviation)
            .field("y", &self.y)
            .field("z", &self.z)
            .finish()
    }
}

impl<T: Dimension + ?Sized> Vector3<T> {
    /// Create a 3 dimensional vector from 3 quantities
    pub fn new(x: Quantity<T>, y: Quantity<T>, z: Quantity<T>) -> Self {
        Self { x, y, z }
    }

    /// Create a zeroed vector
    pub fn zero() -> Self {
        Self {
            x: Quantity {
                dimension: PhantomData,
                units: PhantomData,
                value: 0.0,
            },
            y: Quantity {
                dimension: PhantomData,
                units: PhantomData,
                value: 0.0,
            },
            z: Quantity {
                dimension: PhantomData,
                units: PhantomData,
                value: 0.0,
            },
        }
    }
}

/// Various control signals that the avionics can produce
#[repr(C)]
#[derive(Debug)]
pub enum Control {
    /// An abort signal
    ///
    /// This will stop the system from calling the guidance avionics, and attempt
    /// to enter the vehicle into a recovery or failsafe mode
    ABORT(AbortCause),
    /// A firmware agnostic guidance control signal for the underlying firmware
    /// to translate into flight hardware specific servo movements of pyro channel
    /// fires
    Guidance(Guidance),
    /// A signal to the underlying flight system that the avionics was unable to
    /// compute a valid control signal for the time. The flight system will normally
    /// choose to request another guidance control immediately.
    RecoverableFailure,
}

/// Underlying cause for an abort
///
/// This cause can signal to the abort handler what recovery mode or actions need
/// to be taken and the severity of the abort
#[repr(C)]
#[derive(Debug)]
pub enum AbortCause {
    /// TODO: No abort causes exist currently
    TODO,
}

/// A hardware agnostic guidance signal
#[repr(C)]
#[derive(Debug)]
pub struct Guidance {
    /// Thrust vectoring control
    pub tvc: ThrustVector,
    // TODO: pyro/flags
}

/// A call for thrust vectoring hardware to produce a thrust at the given
/// vector
#[repr(C)]
#[derive(Debug)]
pub struct ThrustVector {
    /// The thrust on the x axis
    pub x: Quantity<angle::Dimension>,
    /// The thrust on the z axis
    pub z: Quantity<angle::Dimension>,
}

/// Hardware agnostic avionics system
///
/// Implementations of this trait should have the [`avionics_harness`] attribute
/// macro preceding them. Alone, this trait has little use.
pub trait Avionics: Debug + Send + Sync {
    /// Produce a control signal given the current sensor values
    ///
    /// This function can be thought of as the control signal generation step
    /// in a control loop
    fn guide(&mut self, sensors: &Sensors) -> Option<Control>;
    // TODO: ABORT HANDLE
    // TODO: CUSTOM SENSORS OR CONTROL STRUCT/ENUM?
}
