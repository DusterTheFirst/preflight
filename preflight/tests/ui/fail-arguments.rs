use preflight::{avionics_harness, Avionics, Control, Sensors};

#[derive(Debug)]
struct Controller;

impl Controller {
    const fn new() -> Self {
        Controller
    }
}

#[avionics_harness(default = "Controller::new()", arg = "ument")]
impl Avionics for Controller {
    fn guide(&mut self, _: &Sensors) -> Control {
        todo!()
    }
}

fn main() {}
