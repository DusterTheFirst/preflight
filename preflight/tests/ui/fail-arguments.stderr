error: Unknown field: `arg`
  --> $DIR/fail-arguments.rs:12:49
   |
12 | #[avionics_harness(default = "Controller::new", arg = "ument")]
   |                                                 ^^^

warning: unused imports: `Avionics`, `Control`, `Sensors`
 --> $DIR/fail-arguments.rs:1:35
  |
1 | use preflight::{avionics_harness, Avionics, Control, Sensors};
  |                                   ^^^^^^^^  ^^^^^^^  ^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
